Run tests on all children images (#1634)
jupyter/docker-stacks: docs/contributing/tests.md

# Image Tests

We greatly appreciate pull requests that extend the automated tests that vet the basic functionality of the Docker images.

## How the Tests Work

> GitHub Action executes `make build-test-all` against pull requests submitted to the `jupyter/docker-stacks` repository.

Reviewer: I believe this is outdated and it looks more like this

```suggestion
A [GitHub Action workflow](https://github.com/jupyter/docker-stacks/blob/master/.github/workflows/docker.yml)
runs the following commands against pull requests submitted to the `jupyter/docker-stacks` repository:

1. ` make -C main build-all-multi` - which builds all the Docker images
2. `make -C main test-all` - which tests the newly created Docker images
```

This `make` command builds and then tests every docker image.

We use `pytest` module to run tests on the image.
`conftest.py` and `pytest.ini` in the `tests` folder define the environment in which tests are run.
More info on `pytest` can be found [here](https://docs.pytest.org/en/latest/contents.html).

> All the actual test files are located in folders like `tests/<somestack>-notebook`.

Author: ```suggestion
All the actual test files are located in folders like `tests/<somestack>-notebook/`.
```

> All the actual test files are located in folders like `tests/<somestack>-notebook/`.

Reviewer: Shall we reword to something like "the actual image-specific tests..."  even though there is a note below

Author: Done, I also think it sounds better

```{note}

> If your test is located in `tests/<somestack>-notebook`, it will be run against `jupyter/<somestack>-notebook` image and against all the images inherited from this image.

Author: ```suggestion
If your test is located in `tests/<somestack>-notebook/`, it will be run against `jupyter/<somestack>-notebook` image and against all the images inherited from this image.
```

> If your test is located in `tests/<somestack>-notebook/`, it will be run against `jupyter/<somestack>-notebook` image and against all the images inherited from this image.

Reviewer: ```suggestion
If your test is located in `tests/<somestack>-notebook/`, it will be run against `jupyter/<somestack>-notebook` image and against all the [images inherited from this image](https://jupyter-docker-stacks.readthedocs.io/en/latest/using/selecting.html#image-relationships. 
```

```

Many tests make use of global [pytest fixtures](https://docs.pytest.org/en/latest/reference/fixtures.html)
defined in the [conftest.py](https://github.com/jupyter/docker-stacks/blob/master/tests/conftest.py) file.

## Unit tests

If you want to run a python script in one of our images, you could add a unit test.
You can do this by creating a `tests/<somestack>-notebook/units/` directory, if it doesn't already exist and put your file there.

> Files in this folder will run automatically when tests are run.

Author: ```suggestion
Files in this folder will be executed in container when tests are run.
```

> Files in this folder will be executed in container when tests are run.

Reviewer: ```suggestion
Files in this folder will be executed in the container when tests are run.
```

> You could see an example for tensorflow package [here](https://github.com/jupyter/docker-stacks/blob/HEAD/tests/tensorflow-notebook/units/unit_tensorflow.py).

Reviewer: ```suggestion
You could see an [example for the TensorFlow package here](https://github.com/jupyter/docker-stacks/blob/HEAD/tests/tensorflow-notebook/units/unit_tensorflow.py).
```

## Contributing New Tests

Please follow the process below to add new tests:

> 1. Add your test code to one of the modules in `<somestack>-notebook/tests/` directory or create a new module.

Author: ```suggestion
1. Add your test code to one of the modules in `tests/<somestack>-notebook/` directory or create a new module.
```

# Image Tests

We greatly appreciate pull requests that extend the automated tests that vet the basic functionality of the Docker images.

## How the Tests Work

A [GitHub Action workflow](https://github.com/jupyter/docker-stacks/blob/master/.github/workflows/docker.yml)
runs the following commands against pull requests submitted to the `jupyter/docker-stacks` repository:

1. `make -C main build-all-multi` - which builds all the Docker images
2. `make -C main test-all` - which tests the newly created Docker images
   This `make` command builds and then tests every docker image.

We use `pytest` module to run tests on the image.
`conftest.py` and `pytest.ini` in the `tests` folder define the environment in which tests are run.
More info on `pytest` can be found [here](https://docs.pytest.org/en/latest/contents.html).

> All the actual test files are located in folders like `tests/<somestack>-notebook/`.

Author: ```suggestion
The actual image-specific test files are located in folders like `tests/<somestack>-notebook/`.
```
